specs/src/recovery.md

```diff
@@ -0,0 +1,525 @@
+
+## Security
+
+As discussed in the intro, the above spec is not yet secure. We have to combine
+2 different mechanisms, parity data and a validator overlay network. Before we
+go over how those mechanisms defend against attacks, we first walk through the
+attacks.
+
+Currently, >1/3 voting power could halt the network, therefore any new
+broadcasting mechanism must be held to same standard. Meaning that if 2/3 of the
+voting power is honest, there should not be attacks that can break the block
+propagation mechanism either.
+
+The only assumptions that we can make are:
+
+**- 1) Nodes are not eclipse attacked** **- 2) 2/3 of the validators are
+participating as described in this spec**
+
+The first assumption means that there exists *a* path of honest nodes connecting
+the broadcaster and every other honest node. The second is self explanatory.
+
+Recall the constraints for recovery:
+
+>- **100% of the Block data MUST be delivered to >2/3 of the validators**
+>- **The time to distribute that data MUST be <= the `ProposalTimeout`**
+
+While pull based gossip and broadcast trees are incredibly powerful tools for
+increasing throughput, it's easy to derive attacks that fail these constraints
+even with the assumptions. First we discuss those attacks, and then we will
+discuss strategies for mitigation.
+
+### Denial of Service
+
+The purpose of the strict diconnection rules are to prevent all scenarios where
+an attacker can flood the network with valid messages. If invalid messages are
+sent, then the attacker is kicked from their address book.
+
+Currently, there are no known gaps. Please comment if you find one.
+
+### Sybil Attacks
+
+#### Pull Based Gossip
+
+Upon requesting data in a standard pull based protocol, faulty peers can simply
+not send requested data. Typically, honest nodes add a timeout, after which they
+will request the data from a different peer. Given our assumption that nodes are
+not being eclipse attacked, nodes will eventually find an honest peer with the
+data.
+
+The problem arises when applications have less time to distribute data than they
+can avoid or remove an attacker. Recovery is such as case. If an attacker wants
+to disrupt a network, it only needs to slow the distribution of data
+sufficiently beyond the `ProposalTimeout`. While the network will eventually
+distribute that data, the throughput will be dramatically decreased until
+connections to malicious peers are removed.
+
+> Side note: its easy to confuse a congested peer as a malicious one. This makes
+> peer scoring incredibly difficult for pull based systems. If timeouts for
+> scoring or addional requests are too agressive, feedback loops emerge.
+> Congested nodes can hit timeouts, resulting in requesting more duplicate data,
+> which causes more congestion.
```

Comment:
Ideally, a peer (and the wider network) can manage it's congestion in a way that we can limit the amount of times that congested nodes behave like malicious nodes (there are also behaviours which are only done by malicious nodes i.e. over requesting data).